lib/std/once.nim

```diff
@@ -0,0 +1,39 @@
+import std/[atomics, locks]
+
+
+type
+  Once* = object    ## A object that ensures a block of code is executed once.
+    finished: Atomic[bool]
+    lock: Lock
+
+proc initOnce*(once: var Once) =
+  ## Initializes a `Once` object.
+  once.finished.store(false)
+  initLock(once.lock)
+
+template once*(cond: Once, body: untyped) =
+  ## Executes a block of code only once (the first time the block is reached).
+  ## It is thread-safe.
+  runnableExamples:
+    var block1: Once
+    var count = 0
+    initOnce(block1)
+
+    for i in 1 .. 10:
+      once(block1):
+        inc count
+
+      # only the first `block1` is executed
+      once(block1):
+        count = 888
+
+    doAssert count == 1
+
+  if not cond.finished.load(moAcquire):
+    withLock cond.lock:
+      # TODO load a value without atomic
+      if not cond.finished.load(moRelaxed):
+        try:
+          body
+        finally:
+          cond.finished.store(true, moRelease)
```

Comment:
I'm wondering whether we should customize whether `cond.finished.store(true, moRelease)` happens in a `finally` block or only after a successful `body` that didn't raise.

there are use cases for both.
eg:
```nim
var db: MyDataBase
proc fn=
  once(Once, retryOnFailure = true):
    db = newMyDataBaseFlaky() # suppose this fails 10% of the times
  # we want to make sure `db` is always initialized here
  db.call() # if `newMyDataBase` raised in another thread, `db` will be nil
```

without `retryOnFailure` flag, how would you emulate `retryOnFailure` in user code calling `std/once` ?